osu.Game/Tests/Visual/Multiplayer/TestMultiplayerClient.cs

```diff
@@ -282,13 +296,78 @@ public override Task ChangeUserMods(IEnumerable<APIMod> newMods)
             return Task.CompletedTask;
         }
 
+        private CancellationTokenSource? countdownFinishSource;
+        private CancellationTokenSource? countdownStopSource;
+        private Task countdownTask = Task.CompletedTask;
+
+        public void FinishCountdown() => countdownFinishSource?.Cancel();
+
         public override async Task SendMatchRequest(MatchUserRequest request)
         {
             Debug.Assert(Room != null);
             Debug.Assert(LocalUser != null);
 
             switch (request)
             {
+                case StartMatchCountdownRequest matchCountdownRequest:
+                    countdownStopSource?.Cancel();
+
+                    var stopSource = countdownStopSource = new CancellationTokenSource();
+                    var finishSource = countdownFinishSource = new CancellationTokenSource();
+                    var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(stopSource.Token, finishSource.Token);
+                    var countdown = new MatchStartCountdown { TimeRemaining = matchCountdownRequest.Duration };
+
+                    Task lastCountdownTask = countdownTask;
+                    countdownTask = start();
+
+                    async Task start()
+                    {
+                        await lastCountdownTask;
+
+                        Schedule(() =>
+                        {
+                            if (stopSource.IsCancellationRequested)
+                                return;
+
+                            Room.Countdown = countdown;
+                            MatchEvent(new CountdownChangedEvent { Countdown = countdown });
+                        });
+
+                        try
+                        {
+                            await Task.Delay(matchCountdownRequest.Duration, cancellationSource.Token).ConfigureAwait(false);
```

Comment:
I still don't understand this. If a stop is requested, it should `return` here right? Not catch and ignore the cancellation?

It seems the only reason this is working is due to the equality check after the schedule, which from a functional perspective is correct, but every time I read this code I am going to think there is an oversight.

Could the nulling of the countdown not be handled in the `catch` instead of in the `StopCountdownRequest`?

Reply:
I explained this here: https://github.com/ppy/osu-server-spectator/pull/109#discussion_r831891849

We need to run all of the following code regardless of whether or not the countdown was cancelled.

Reply:
Aha, makes more sense. Can the same comment be added here in the `catch` block?